Documentation/backends.md

```diff
@@ -164,3 +164,18 @@ Troubleshooting
 * `ip xfrm policy` can be used to show the installed policies. Flannel installs three policies for each host it connects to. 
 
 Flannel will not restore policies that are manually deleted (unless flannel is restarted). It will also not delete stale policies on startup. They can be removed by rebooting your host or by removing all ipsec state with `ip xfrm state flush && ip xfrm policy flush` and restarting flannel.
+
+### WireGuard
+
+Use in-kernel [WireGuard](https://www.wireguard.com) to encapsulate and encrypt the packets.
+
+Type:
+* `Type` (string): `wireguard`
+* `PSK` (string): Optional. The pre shared key to use. Use `wg genpsk` to generate a key.
+* `ListenPort` (int): Optional. The udp port to listen on. Default is `51820`.
+* `ListenPortV6` (int): Optional. The udp port to listen on for ipv6. Default is `51821`.
```

Comment:
Why is this two different wg interfaces, with different listen addresses? `wg` should support some v4 and some v6 peers just fine, and the `ListenPort` should support dualstack just fine.

Reply:
I made this decision because I want to ensure all IPv6 traffic is even routed between the nodes over IPv6. As far as I know wireguard handles public keys from peers unique. So it should not be possible to have one peer with IPv4 endpoint and one peer with IPv6 endpoint with the same public key. I know I can add a peer with IPv4 endpoint and route IPv4/IPv6 subnets over it. But this would not allow us to run this backend in IPv6 only mode. Am I wrong?

Reply:
Wireguard doesn't really care whether it receives a packet over IPv4 or IPv6, and from which address it's originated. As long as it's properly signed, it's considered valid (and the endpoint in the peer list is updated).

This is especially true if you have a dualstacked machine in a datacenter, providing a VPN endpoint, and have "roadwarrior" laptops connecting to it. In that case, there's not even a stable endpoint preconfigured in the server config at all (and the server can't talk to the clients unless they initially established the connection).

On a wireguard interface, you can add both IPv4/IPv6 addresses/routes and send packets, no matter which address family the underlay network uses. I think that's a strength - let's say a node only has outside IPv4 connectivity - you could still use IPv6 addressing internally.

> I made this decision because I want to ensure all IPv6 traffic is even routed between the nodes over IPv6

I'm not sure I fully understand why this would this be desirable.
It requires a separate network interface, a separate list of peers, a distinct udp listen port. Also, how would it behave if one node would only have a public IPv4 address, but no IPv6 address? That traffic would just get blackholed?

Reply:
> Wireguard doesn't really care whether it receives a packet over IPv4 or IPv6, and from which address it's originated. As long as it's properly signed, it's considered valid (and the endpoint in the peer list is updated).
> 
> This is especially true if you have a dualstacked machine in a datacenter, providing a VPN endpoint, and have "roadwarrior" laptops connecting to it. In that case, there's not even a stable endpoint preconfigured in the server config at all (and the server can't talk to the clients unless they initially established the connection).
> 
> On a wireguard interface, you can add both IPv4/IPv6 addresses/routes and send packets, no matter which address family the underlay network uses. I think that's a strength - let's say a node only has outside IPv4 connectivity - you could still use IPv6 addressing internally.

Well, looks like we share the same passion about wireguard.. :)

> 
> > I made this decision because I want to ensure all IPv6 traffic is even routed between the nodes over IPv6
> 
> I'm not sure I fully understand why this would this be desirable. It requires a separate network interface, a separate list of peers, a distinct udp listen port.

I'm not happy about this overhead either. But to ensure the inter-host communication is based on IPv6 there is no other way.

My suggestion is to add parameters to wireguard configuration to enable v4 in v6 or v6 in v4 for inter-host communication in dual-stack mode. If one of these options is enabled no second device will be created. But I still think the default should be not to force any encapsulation.

> Also, how would it behave if one node would only have a public IPv4 address, but no IPv6 address? That traffic would just get blackholed?

Flannel requires a public v4 and v6 address for dual-stack support.

Reply:
> I'm not happy about this overhead either. But to ensure the inter-host communication is based on IPv6 there is no other way.

This is only true if we use the same private key for v6 traffic. If we generate separate private key for v6 traffic we have a unique public key for the v6 endpoint and can handle this with one device. I will implement the required changes tomorrow.

Reply:
> My suggestion is to add parameters to wireguard configuration to enable v4 in v6 or v6 in v4 for inter-host communication in dual-stack mode. If one of these options is enabled no second device will be created. But I still think the default should be not to force any encapsulation.

There is encapsulation in both cases - you either send the source and destination ipv4 or ipv6 packets in the udp packet payload.

> Flannel requires a public v4 and v6 address for dual-stack support.

Eww, I wasn't aware of that. In that case, wouldn't it be preferrable to use the public IPv6 addresses of the nodes (if available), even to transport IPv4 internal traffic?

IPv4 address shortage isn't going to go away - if we can stop requiring public IPv4 adresses on every worker nodes, that'd help.

Reply:
> > My suggestion is to add parameters to wireguard configuration to enable v4 in v6 or v6 in v4 for inter-host communication in dual-stack mode. If one of these options is enabled no second device will be created. But I still think the default should be not to force any encapsulation.
> 
> There is encapsulation in both cases - you either send the source and destination ipv4 or ipv6 packets in the udp packet payload.

If the udp packet is always send to a v4 endpoint there will be no v6 traffic in inter-host communication. That what I meant with encapsulation.

> > Flannel requires a public v4 and v6 address for dual-stack support.
> 
> Eww, I wasn't aware of that. In that case, wouldn't it be preferrable to use the public IPv6 addresses of the nodes (if available), even to transport IPv4 internal traffic?
> 
> IPv4 address shortage isn't going to go away - if we can stop requiring public IPv4 adresses on every worker nodes, that'd help.

It's a hard guess what will be the preferable configuration for all flannel users. I think the best experience for flannel users would be to make this configurable.

Reply:
Since I don't have in time in the next days, I would like to send another PR in the next weeks, which will allow to configure the routing in inter-host communication in dual-stack mode.

Reply:
"Flannel requires a public v4 and v6 address for dual-stack support."

That seems a misconception; With the patches i did for the flannel extension backend && k3s to do dual-stack i was happily running a multi-node k3s with one node having both public ipv4 and ipv6 and the other having only a private ipv4 address.. (And also using a local ipv6 range inside the cluster)..

Fundamentally what you need with the extension backend + wireguard and, i would assume, this backend as well is that you need to be able to setup a full mesh (fancier routing like e.g. kilo can i would guess it out of scope for flannel?).

 In other words every pair of nodes need to be able to setup a wireguard tunnel between them; Which can happen as long as one of the nodes can connect to the initial configured endpoint of the other. e.g. in a setup where node A has public ip(s) but B doesn't, it doesn't matter that A can't connect to B, the wireguard tunnel will be setup once B connects to A. 

Unfortunately wireguard doesn't support configuring multiple (initial) endpoints. So the whole challenge is selecting the right initial endpoints configuration on each side (again as long as one side gets it right things will work) or alternatively monitoring and trying different options if the tunnel doesn't end up getting set up.

Fwiw for extra fun; If one side you setup a ipv6 address as the remote endpoint and on the other side an ipv4 one initially whether you end up tunneling over v4 or v6 just depends on the one that initiates successfully first, wireguard really doesn't care about the underlying address family.